metrics/src/main/java/org/aerogear/mobile/metrics/MetricsService.java

```diff
@@ -10,67 +9,58 @@
 import org.aerogear.mobile.core.http.HttpResponse;
 import org.aerogear.mobile.core.http.HttpServiceModule;
 import org.aerogear.mobile.core.logging.Logger;
+import org.aerogear.mobile.metrics.interfaces.MetricsProvider;
+import org.aerogear.mobile.metrics.providers.DefaultMetricsProvider;
+import org.json.JSONArray;
 import org.json.JSONException;
 import org.json.JSONObject;
 
-import java.util.UUID;
-
 public class MetricsService implements ServiceModule {
     public final static String STORAGE_NAME = "org.aerogear.mobile.metrics";
     public final static String STORAGE_KEY = "metrics-sdk-installation-id";
 
     private final static String MODULE_NAME = "metrics";
-    private final static String TAG = "AEROGEAR/METRICS";
+    public final static String TAG = "AEROGEAR/METRICS";
 
     private HttpServiceModule httpService;
     private Logger logger;
 
-    private String appVersion = null;
     private String metricsUrl = null;
 
-    /**
-     * Get or create the client ID that identifies a device as long as the user doesn't
-     * reinstall the app or delete the app storage. A random UUID is created and stored in the
-     * application shared preferences.
-     *
-     * Can be overridden to provide a different implementation for identification.
-     *
-     * @param context Android app context
-     * @return String Client ID
-     */
-    protected String getOrCreateClientId(final Context context) {
-        final SharedPreferences preferences = context
-            .getSharedPreferences(STORAGE_NAME, Context.MODE_PRIVATE);
-
-        String clientId = preferences.getString(STORAGE_KEY, null);
-        if (clientId == null) {
-            clientId = UUID.randomUUID().toString();
-
-            logger.info(TAG, "Generated a new client ID: " + clientId);
-
-            SharedPreferences.Editor editor = preferences.edit();
-            editor.putString(STORAGE_KEY, clientId);
-            editor.commit();
+    // Copy all properties from the source object and add it to the top level
+    // of the root object
+    private void addToRoot(final JSONObject root, final JSONObject source) throws JSONException {
+        JSONArray names = source.names();
+        for (int i = 0; i < names.length(); ++i) {
+            String name = names.getString(i);
+            root.put(name, source.get(name));
         }
+    }
 
-        return clientId;
+    private void aggregateSingleProvider(
+        final MetricsProvider provider,
+        final JSONObject target,
+        final Context context)
+        throws JSONException {
+
+        String namespace = provider.namespace();
+        JSONObject metrics = provider.metrics(context);
+
+        // If the namespace of a provider is null we add it's properties
+        // to the root object
+        if (namespace == null || namespace.length() == 0) {
```

Comment:
Not sure if we should allow this case and threat that as error or do it this way.
I have no preference for this but flagging that as possible problem.

Reply:
I added this mostly to support the current format (without nesting). But i can remove it if we decide to enforce namespacing.

Reply:
For me, having a null or empty namespace would look like an error. When would this condition be met?

Reply:
If you don't want to namespace your data and just append it to the root of the object

Reply:
IMO all providers should have a namespace. Even the default ones could have it, for consistency and to eliminate this null/empty namespace approach.